File: DotNetOnion/Cells/CellAuthChallenge.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Text;
using DotNetOnion.Helpers;

namespace DotNetOnion.Cells
{
    public class CellAuthChallenge : Cell
    {
        public byte[] Challenge { get; set; }
        public List<ushort> Methods { get; set; } = new List<ushort>();
        public override byte Command => 130;

        public override void Deserialize(BinaryReader reader)
        {
            Challenge = reader.ReadBytes(32);
            var len = reader.ReadUInt16BigEndian();
            for (int i = 0; i < len; i++)
            {
                Methods.Add(reader.ReadUInt16BigEndian());
            }
        }

        public override void Serialize(BinaryWriter writer)
        {
            writer.Write(Challenge);
            writer.WriteUInt16BigEndian((ushort)Methods.Count);
            foreach (var method in Methods)
            {
                writer.WriteUInt16BigEndian(method);
            }
        }
    }

}

File: NOnion.Core/Cells/CellAuthChallenge.fs
namespace NOnion.Core.Cells

open System.IO

open FSharpx.Collections

open NOnion.Core
open NOnion.Core.Utility.BinaryIO

type CellAuthChallenge =
    {
        Challenge: array<byte>
        Methods: seq<uint16>
    }

    static member Deserialize(reader: BinaryReader) =

        let rec readMethod methods n =
            if n = 0 then
                methods
            else
                readMethod(methods @ [ ReadBigEndianUInt16 reader ]) (n - 1)

        let challenge = reader.ReadBytes Constants.ChallangeLength
        let methodsCount = ReadBigEndianUInt16 reader |> int
        let methods = readMethod [] methodsCount

        {
            Challenge = challenge
            Methods = methods
        }
        :> ICell

    interface ICell with

        member __.Command = 130uy

        member self.Serialize writer =

            let rec writeMethods(methods: seq<uint16>) =
                match Seq.tryHeadTail methods with
                | None -> ()
                | Some(method, nextMethods) ->
                    WriteUInt16BigEndian writer method
                    nextMethods |> writeMethods

            writer.Write self.Challenge
            self.Methods |> Seq.length |> uint16 |> WriteUInt16BigEndian writer
            writeMethods self.Methods
